src/compiler/scanner.h

```diff
@@ -129,15 +129,124 @@ static inline bool is_binary_digit(int c) {
   return c == '0' || c == '1';
 }
 
-static inline bool is_identifier_start(int c) {
-  ASSERT(c >= 0);
-  return c == LSP_SELECTION_MARKER || is_letter(c) || (c == '_');
-}
+class IdentifierValidator {
+ public:
+  // Returns whether the given character is part of the started identifier.
+  // If this is the first character also checks that it can start an
+  // identifier (unless `disable_start_check` was called before).
+   // The given `peek` function is only invoked if 'c' is equal to '-'.
+  bool check_next_char(uint8 c, const std::function<uint8 ()>& peek) {
+    if (c == 0) return false;
+    if (at_start_) {
+      at_start_ = false;
+      return is_identifier_start(c);
+    }
+    if (c == '-') {
+      uint8 next = peek();
+      if (next == 0) return false;
+      return is_identifier_part_no_dash(next);
+    }
+    return is_identifier_part_no_dash(c);
+  }
 
-static inline bool is_identifier_part(int c) {
-  ASSERT(c >= 0);
-  return c == LSP_SELECTION_MARKER || is_letter(c) || is_decimal_digit(c) || (c == '_');
-}
+  // Sets the internal state so it doesn't check for an identifier start.
+  // This is useful when moving backwards (as is done for finding the
+  // completion prefix.
+  void disable_start_check() {
+    at_start_ = false;
+  }
+
+  static bool is_identifier_start(int c) {
+    ASSERT(c >= 0);
+    return c == LSP_SELECTION_MARKER || is_letter(c) || (c == '_');
+  }
+
+  static const uint8* canonicalize(const uint8* identifier, int len) {
+    if (len < 3) return identifier;
+
+    // Check whether the identifier has a '_'. If yes, convert it to '-'.
+    // We only change '_'s if they are surrounded by characters.
+    bool needs_canonicalization = false;
+    for (int i = 1; i < len - 1; i++) {
+      uint8 c = identifier[i];
+      if (c != '_') continue;
+
+      uint8 previous = identifier[i - 1];
+      uint8 next = identifier[i + 1];
+      if (previous == '-' || previous == '_') continue;
+      if (next == '-' || next == '_') continue;
+      needs_canonicalization = true;
+      break;
+    }
+    if (!needs_canonicalization) return identifier;
+
+    uint8* copy = reinterpret_cast<uint8*>(malloc(len + 1));
+    memcpy(copy, identifier, len);
+    // The terminating character might not have been present in the
```

Comment:
I don't understand this comment. You only get the terminating \0 if the string needs canonicalization, so this seems a bit brittle. Where do you use this from where the input has no terminating \0, but you need the output to have it?